PR resolved

dommrogers opened the pull request:
[Better-Stacking.zip](https://github.com/ds5678/Better-Stacking/files/10296239/Better-Stacking.zip)

dommrogers committed:
Merge pull request #1 from ds5678/master

WIP changes for MelonLoader 6 and TLD 2.04

ds5678 committed:
Revert "== null instead of is null and other small refactorings"

This reverts commit 545d755fb7c8c819aa6166ad61f032898a5580b0.

dommrogers committed:
rework harmonyPatch & fix stacking bug

dommrogers committed:
TODO - fix removal of gearToAdd to avoid duplication

dommrogers committed:
Rework of "MergeIntoStack"

Work around parent method "InstantiateItemInPlayerInventory" causing item dupes for console & new gear.

dommrogers committed:
Significant rework for TLD 2.06 & fix console stacking

dommrogers committed:
Merge branch 'tld2.06'

dommrogers committed:
Restore commit "== null instead of is null and other small refactorings"

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -72,13 +57,60 @@ internal static bool CanBeMerged(GearItem target, GearItem item)
+
+        // check for console added items
+        // condition is always 1 as this gets changed later even when specified
+        if (normalizedCondition == 1 && uConsole.IsOn())
```
`normalizedCondition == 1` is probably not necessary and can be removed.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -1,13 +1,14 @@
 {
-    private static readonly string[] STACK_MERGE =
-    {
+    private static readonly string[] STACK_MERGE = {
```
Put the curly bracket where it was before.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,98 @@
+
+        internal static bool postfix_track = false;
+        internal static float postfix_condition = 0;
+        internal static GearItem postfix_stack = new GearItem();
+        internal static GearItem postfix_geartoadd = new GearItem();
```
Do not call new() on GearItem. Make this nullable

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,98 @@
-        }
+    {
+
+        internal static bool postfix_track = false;
+        internal static float postfix_condition = 0;
+        internal static GearItem postfix_stack = new GearItem();
+        internal static GearItem postfix_geartoadd = new GearItem();
+        internal static float postfix_constraint = 0;
```
Convert these to `{ get; private set }` properties for code encapsulation.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,98 @@
+                        return m;
+                    }
+                }
+                MelonLogger.Msg("PlayerManager.TryAddToExistingStackable not found for patch.");
```
Warn instead of Msg

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,98 @@
+            internal static bool Prefix(ref GearItem gearToAdd, float normalizedCondition, int numUnits, out GearItem existingGearItem)
+            {
+
+                existingGearItem = new GearItem();
```
As above, don't create new Unity components from managed code. Make it nullable.

ds5678 replied:
I'm fairly sure this line can be removed entirely.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,98 @@
+
+                GearItem targetStack = GameManager.GetInventoryComponent().GetClosestMatchStackable(gearToAdd.GearItemData, normalizedCondition);
+
+                if (!Implementation.CanBeMerged(targetStack, gearToAdd))
```
Move CanBeMerged into this file. It doesn't belong in Implementation. I was a worse programmer when I wrote it.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,98 @@
+
+                }
+
+                // reset the static values to avoid any conflicts
+                postfix_track = false;
+                postfix_condition = 0;
+                postfix_stack = new GearItem();
+                postfix_geartoadd = new GearItem();
+                postfix_constraint = 0;
```
Make a Reset method

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -72,13 +57,60 @@ internal static bool CanBeMerged(GearItem target, GearItem item)
     internal static void Log(string message) => MelonLogger.Msg(message);
     internal static void Log(string message, params object[] parameters) => MelonLogger.Msg(message, parameters);
```
Delete these. They access obsolete methods. The logger instance is a MelonMod property.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -1,13 +1,14 @@
-using MelonLoader;
+﻿using System.Linq;
 using UnityEngine;
-
+using MelonLoader;
+using Il2Cpp;
```
If you right-click, vs has an option to sort usings in the current file.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -72,13 +57,60 @@ internal static bool CanBeMerged(GearItem target, GearItem item)
+            Patches.postfix_constraint = targetStack.m_StackableItem.m_StackConditionDifferenceConstraint;
+        }
+
+        /*
+         convince the game logic these stacks can merge between 0% -> 100% condition
+         the game keeps the existing stack condition %
+        (avoids having to destroy anything, just let the game do it all)
+        */
```
Use line comments instead of block comments

ds5678 reviewed (changes requested):
I took a look over your pr and noticed a few things.

dommrogers committed:
Code adjustments suggested by ds5678

Had to use { get; set } as they are used outside of the Patches class (Patches.cs 11 to 15)

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -1,12 +1,13 @@
+﻿using Il2Cpp;
 using MelonLoader;
 using UnityEngine;
-
```
It's generally good to have a line between the usings and namespace.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -45,40 +46,65 @@ public override void OnInitializeMelon()
+
+
+
+    internal static void MergeIntoStack(float normalizedCondition, int numUnits, GearItem targetStack, GearItem gearToAdd)
```
Move this method out of Implementation. It doesn't belong here.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -94,79 +120,18 @@ internal static void SplitStack(GearItem gearItem)
         catch (System.Exception e)
         {
-            Log("Failed to split stack of {0}: {1}.", gearItem.name, e);
+            MelonLogger.Warning("Failed to split stack of {0}: {1}.", gearItem.name, e);
```
The `MelonLogger` static methods are obsolete. You need to use the `Implementation.LoggerInstance`.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -94,79 +120,18 @@ internal static void SplitStack(GearItem gearItem)
+
+
+
     private static void SplitStackWithException(GearItem gearItem)
```
Move this too.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -94,79 +120,18 @@ internal static void SplitStack(GearItem gearItem)
+
+
+
+
```
Only one line between methods.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -94,79 +120,18 @@ internal static void SplitStack(GearItem gearItem)
         }
     }
 
     internal static bool UseDefaultStacking(GearItem gearItem)
```
This can be moved to Patches. Implementation needs almost nothing.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,127 @@
+        {
+            static MethodBase? TargetMethod()
+            {
+                MethodInfo[] methods = typeof(PlayerManager).GetMethods();
+                foreach (MethodInfo m in methods)
+                {
+                    if (m.Name == nameof(PlayerManager.TryAddToExistingStackable) && m.ReturnType == typeof(bool) && !m.IsGenericMethod && m.GetParameters().Length == 4)
+                    {
+                        return m;
+                    }
+                }
```
Not a requirement, but this code could be implemented cleaner with System.Linq

ds5678 replied:
For example:
```cs
            static MethodBase? TargetMethod()
            {
                MethodInfo? targetMethod = typeof(PlayerManager)
                    .GetMethods()
                    .FirstOrDefault(m => m.Name == nameof(PlayerManager.TryAddToExistingStackable)
                        && m.ReturnType == typeof(bool)
                        && !m.IsGenericMethod && m.GetParameters().Length == 4);
                if (targetMethod is null)
                {
                    Implementation.LogWarning("PlayerManager.TryAddToExistingStackable not found for patch.");
                }

                return targetMethod;
            }
```

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,127 @@
+        {
+            private static void Postfix()
+            {
+                // are we tracking a postfix patch ?
```
This kind of thing should be an xml comment on `PostFixTrack`.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,127 @@
+            }
+
+            return true;
+        }
```
I think this whole thing can be removed since we don't do anything with flares.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,127 @@
+            }
+        }
+
+        internal static bool CanBeMerged(GearItem target, GearItem item)
```
Since we allow null input, the parameter types should be marked as nullable.

Also, it's good programming to mark this method as null checking.
```cs
private static bool CanBeMerged([NotNullWhen(true)] GearItem? target, [NotNullWhen(true)] GearItem? item)
```

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,127 @@
+
+        internal static bool CanBeMerged(GearItem target, GearItem item)
+        {
+            return target != null && item != null && CanBeMerged(target.m_FlareItem, item.m_FlareItem);
```
The flare item check can probably be removed.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -45,40 +46,65 @@ public override void OnInitializeMelon()
         //MakeStackable("GEAR_SharpeningStone");
         //MakeStackable("GEAR_SimpleTools");
         //MakeStackable("GEAR_HighQualityTools");
```
These comments can be removed.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -1,12 +1,13 @@
 internal class Implementation : MelonMod
 {
-    private static readonly string[] STACK_MERGE =
+    private static readonly string[] STACK_MERGE =
```
The whole `StackMerge` field can be moved.

ds5678 commented on BetterStacking/Implementation.cs:
```diff
@@ -1,12 +1,13 @@
 using UnityEngine;
-
 namespace BetterStacking;
 
+
+
```
3 empty lines is a bit much. 😄

ds5678 reviewed (changes requested):
Starting to look much better!

dommrogers committed:
Code adjustments suggested by ds5678 #2

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,222 @@
+        "GEAR_WoodMatches",
+        };
+
+        internal static bool PostFixTrack { get; private set; } = false;
+
+        internal static float PostfixCondition { get; private set; } = 0;
+
+        internal static GearItem? PostfixStack { get; private set; } = null;
+
+        internal static GearItem? PostfixGearToAdd { get; private set; } = null;
+
+        internal static float PostfixConstraint { get; private set; } = 0;
```
I think these can be made private.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,222 @@
+            }
+        }
+
+        internal static bool UseDefaultStacking(GearItem gearItem)
```
`GearItem? gearItem`

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,222 @@
+            targetStack.CurrentHP = targetCondition * targetStack.m_GearItemData.m_MaxHP;
+        }
+
+        internal static void SplitStack(GearItem gearItem)
```
`GearItem? gearItem`

ds5678 reviewed (changes requested):
Very close. One more round of changes and I think everything should be good.

dommrogers committed:
remove unused SplitStack methods

dommrogers committed:
Change static variables to be private

dommrogers committed:
rework TargetMethod to be better

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,191 @@
+                return targetMethod;
+            }
+
+            internal static bool Prefix(ref GearItem gearToAdd, float normalizedCondition, int numUnits, ref GearItem existingGearItem)
```
I'd like a comment (ideally xml) saying that `existingGearItem` is an out parameter on the original method. That's extremely important information.

Xml comments can be created in Visual Studio by typing 3 forward slashes above something, ie `///`.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,191 @@
+            {
+                if (normalizedCondition == 0)
+                {
+                    return false;
```
Anytime you return false on an original method with out parameters, you have to set all its out parameters. Otherwise, you'll get undefined behavior from the game when it tries to access the uninitialized out parameter.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,191 @@
+                {
+                    MergeIntoStack(normalizedCondition, numUnits, targetStack, gearToAdd);
+                    existingGearItem = targetStack;
+                    return true;
```
Is your intention not to run the original method here? If so, this should be `return false;`.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,191 @@
+
+                if (!CanBeMerged(targetStack, gearToAdd))
+                {
+                    return false;
```
Why is the original method being skipped in this scenario? I feel like your actual intention is to allow the original to run.

If it actually does need skipped, you need to set the out parameter.

ds5678 commented on BetterStacking/Patches.cs:
```diff
@@ -1,145 +1,191 @@
+                    return true;
+                }
+
+                return false;
```
Same as the rest of the method. I'm confused about "skipping vs not skipping" and ensuring that out parameters get set whenever the original is not run.

ds5678 reviewed (changes requested):
Unfortunately I noticed some potential issues in the patch of `TryAddToExistingStackable`.
* I'm finding it difficult to understand why the method is being skipped in certain situations. Some code comments would really help to explain the reasoning behind the decisions.
* `existingGearItem` needs to be set whenever the original is skipped.

dommrogers committed:
fixed bad skipping logic & added comments

Tested with various items, picking up, adding via console, taking from containers. All seemed to work fine

ds5678 commented:
As said in Discord, I want:

* A bump to version 5.0.0
* Continuous integration similar to https://github.com/ds5678/SeasonedInterloping/tree/main/.github

dommrogers committed:
bump version to 5.0.0

dommrogers committed:
add github CI workflow

dommrogers committed:
add build.yml to project solution

dommrogers committed:
Replace references with NuGet package

dommrogers committed:
fix null deference warnings

ds5678 commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,34 @@
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    branches: [ "main" ]
```
The branch on this repository is "master"

dommrogers committed:
fix build branch

dommrogers committed:
update .csproj to be inline with ModTemplate
